actions/rules/action_field_set.go

```diff
@@ -0,0 +1,105 @@
+package rules
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/convert"
+	"github.com/fabric8-services/fabric8-wit/workitem"
+)
+
+// ActionFieldSet takes a configuration JSON object that has field
+// names as the keys and a value as the argument. It updates the
+// given ChangeDetector and sets the Field[key] value to the
+// values given. Note that this only works on WorkItems.
+type ActionFieldSet struct {
+	Db     application.DB
+	Ctx    context.Context
+	UserID *uuid.UUID
+}
+
+// make sure the rule is implementing the interface.
+var _ Action = ActionFieldSet{}
+
+func (act ActionFieldSet) storeWorkItem(wi *workitem.WorkItem) (*workitem.WorkItem, error) {
+	if act.Ctx == nil {
+		return nil, errs.New("context is nil")
+	}
+	if act.Db == nil {
+		return nil, errs.New("database is nil")
+	}
+	if act.UserID == nil {
+		return nil, errs.New("userID is nil")
+	}
+	var storeResultWorkItem *workitem.WorkItem
+	err := application.Transactional(act.Db, func(appl application.Application) error {
+		var err error
+		storeResultWorkItem, err = appl.WorkItems().Save(act.Ctx, wi.SpaceID, *wi, *act.UserID)
+		if err != nil {
+			return errs.Wrap(err, "error updating work item")
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return storeResultWorkItem, nil
+}
+
+// OnChange executes the action rule.
+func (act ActionFieldSet) OnChange(newContext convert.ChangeDetector, contextChanges []convert.Change, configuration string, actionChanges *[]convert.Change) (convert.ChangeDetector, []convert.Change, error) {
+	// check if the newContext is a WorkItem, fail otherwise.
+	wiContext, ok := newContext.(workitem.WorkItem)
+	if !ok {
+		return nil, nil, errs.New("given context is not a WorkItem: " + reflect.TypeOf(newContext).String())
+	}
+	// deserialize the config JSON.
+	var rawType map[string]interface{}
+	err := json.Unmarshal([]byte(configuration), &rawType)
+	if err != nil {
+		return nil, nil, errs.Wrap(err, "failed to unmarshall from action configuration to a map: "+configuration)
+	}
+	// load WIT.
+	wit, err := act.Db.WorkItemTypes().Load(act.Ctx, wiContext.Type)
+	if err != nil {
+		return nil, nil, errs.Wrap(err, "error loading work item type: "+err.Error())
+	}
+	// iterate over the fields.
+	var convertChanges []convert.Change
+	for k, v := range rawType {
+		if wiContext.Fields[k] != v {
+			fieldType, ok := wit.Fields[k]
+			if !ok {
+				return nil, nil, errs.New("unknown field name: " + k)
+			}
+			convertChanges = append(convertChanges, convert.Change{
```

Comment:
Why create a `convertChanges []convert.Change` when we already have it in function parameters?

https://github.com/fabric8-services/fabric8-wit/pull/2154/files#diff-d3cb7a6ecaf58886b81f275842073ddbR55

Reply:
Good find!! This is actually a bug, the convertChange needs to stack to allow for multiple matching rules with a unified change set available after the last rule executed. This works for the StateToMetaState rule, but had this bug in this rule. Fixed it and added a test.

Reply:
#45331a7